server/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java

```diff
@@ -291,14 +364,21 @@ public ClusterState applyStartedShards(ClusterState clusterState, List<ShardRout
     }
 
     private ClusterState executeClusterStateUpdateTask(ClusterState state, Runnable runnable) {
-        ClusterState[] result = new ClusterState[1];
+        when(clusterService.state()).thenReturn(state);
+        CompletableFuture<ClusterState> clusterStateResult = new CompletableFuture<>();
         doAnswer(invocationOnMock -> {
             ClusterStateUpdateTask task = (ClusterStateUpdateTask)invocationOnMock.getArguments()[1];
-            result[0] = task.execute(state);
+            assertTrue(clusterStateResult.complete(task.execute(state)));
```

Comment:
I changed this to allow us to get the result of an asynchronously completing action triggered by the `runnable` and added the assertion to make sure we're only ever handling a single state update here.